File: apps/ios/App/FeedbackSheet.swift
import Domain
import SwiftUI
import UIKit

/**
 Send feedback — ported from `apps/web/src/ui/BugReport.tsx`.

 The "Feedback" entry in the More sheet and the side nav was wired to a closure
 that closed the sheet and did nothing else, on BOTH platforms: a visible
 control that lied, and the app's only error-report channel.

 Two things are deliberately not web's:

  * **Every string is translated.** Web's modal is hardcoded English end to end
    — 30-odd strings, in an app that ships in three languages. The catalogue
    entries added for this port are there for web to adopt.
  * **The area picker is a chip grid, not a `<select>`.** Fourteen options in a
    native picker is a wheel; chips show the whole vocabulary at once, which is
    what the web `<select>` effectively does on a desktop.

 What IS web's, exactly: the default-on log toggle, the promise about what is
 captured, the bug/suggestion split (and with it the severity row, which a
 suggestion does not get), and the thank-you screen including the beta-tester
 reward copy.

 Mirrors Android's FeedbackSheet.
 */
struct FeedbackSheet: View {
    let route: String
    let online: Bool
    let onClose: () -> Void

    @State private var viewModel = FeedbackViewModel()

    var body: some View {
        Group {
            if viewModel.done {
                donePanel
            } else {
                form
            }
        }
    }

    // MARK: - form

    private var form: some View {
        @Bindable var vm = viewModel
        return VStack(alignment: .leading, spacing: 12) {
            SanvyaH2(S.Feedback.title)
            SanvyaMuted(S.Feedback.intro, style: SanvyaType.body.resized(12))

            HStack(spacing: 6) {
                SanvyaChip(S.Feedback.kindBug, isActive: viewModel.isBug) {
                    vm.kind = "bug"
                    viewModel.clearError()
                }
                SanvyaChip(S.Feedback.kindSuggestion, isActive: !viewModel.isBug) {
                    vm.kind = "suggestion"
                    viewModel.clearError()
                }
                Spacer(minLength: 0)
            }

            // A suggestion has no severity — web hides the whole row, and the
            // repository writes nil, so the triage queue is not sorted by an
            // urgency nobody chose.
            if viewModel.isBug {
                SanvyaMuted(S.Feedback.severityLabel, style: SanvyaType.body.resized(12))
                chipFlow(feedbackSeverities, selected: viewModel.severity) { severity in
                    vm.severity = severity
                } label: { feedbackLabel(feedbackSeverityKey($0)) }
            }

            SanvyaMuted(S.Feedback.areaLabel, style: SanvyaType.body.resized(12))
            chipFlow(feedbackAreas, selected: viewModel.area) { area in
                // Tapping the chosen one again clears it: web's select has an
                // empty first option and this has no equivalent row to offer.
                vm.area = viewModel.area == area ? "" : area
            } label: { feedbackLabel(feedbackAreaKey($0)) }

            SanvyaInput(text: $vm.title, placeholder: S.Feedback.titlePlaceholder)
            TextField(
                viewModel.isBug ? S.Feedback.bugPlaceholder : S.Feedback.suggestionPlaceholder,
                text: $vm.descriptionText,
                axis: .vertical
            )
            .lineLimit(4...8)
            .sanvyaStyle(SanvyaType.body)
            .foregroundStyle(Color.text)
            .padding(.horizontal, 14)
            .padding(.vertical, 11)
            .background(Color.surface)
            .clipShape(RoundedRectangle(cornerRadius: SanvyaRadius.radiusSm, style: .continuous))
            .overlay {
                RoundedRectangle(cornerRadius: SanvyaRadius.radiusSm, style: .continuous)
                    .strokeBorder(Color.border, lineWidth: 1)
            }
            .onChange(of: viewModel.descriptionText) { _, _ in viewModel.clearError() }

            Toggle(isOn: $vm.includeLog) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(S.Feedback.includeLog)
                        .sanvyaStyle(SanvyaType.body.resized(13))
                        .foregroundStyle(Color.text)
                    SanvyaMuted(S.Feedback.includeLogHint, style: SanvyaType.body.resized(11))
                }
            }
            .tint(Color.accent)

            SanvyaMuted(S.Feedback.autoIncluded, style: SanvyaType.body.resized(11))

            if let key = viewModel.errorKey {
                Text(feedbackLabel(key))
                    .sanvyaStyle(SanvyaType.body.resized(13))
                    .foregroundStyle(Color.negative)
            }

            HStack(spacing: 8) {
                Spacer(minLength: 0)
                SanvyaButton(ghost: true, action: { onClose(); viewModel.reset() }) {
                    Text(S.Feedback.cancel)
                }
                SanvyaButton(action: send) {
                    Text(sendLabel)
                }
                .disabled(!viewModel.canSubmit)
            }
        }
    }

    private var sendLabel: String {
        if viewModel.busy { return S.Feedback.sending }
        return viewModel.isBug ? S.Feedback.sendBug : S.Feedback.sendSuggestion
    }

    private func send() {
        let screen = UIScreen.main.bounds.size
        viewModel.submit(
            route: route,
            // Web stores `navigator.userAgent`; the honest equivalent is the
            // device and OS, which is what that string is read FOR.
            userAgent: "iOS \(UIDevice.current.systemVersion); \(UIDevice.current.model)",
            viewport: "\(Int(screen.width))x\(Int(screen.height))",
            online: online
        )
    }

    // MARK: - done

    /// Web's post-submit panel, reward copy and all.
    private var donePanel: some View {
        VStack(spacing: 10) {
            Text(verbatim: "🙏").font(.system(size: 34))
            SanvyaH2(viewModel.isBug ? S.Feedback.thanksBug : S.Feedback.thanksSuggestion)
            SanvyaMuted(
                viewModel.isBug ? S.Feedback.thanksBugBody : S.Feedback.thanksSuggestionBody,
                style: SanvyaType.body.resized(14)
            )
            .multilineTextAlignment(.center)
            HStack(spacing: 8) {
                SanvyaButton(ghost: true, action: { viewModel.reset() }) {
                    Text(S.Feedback.sendAnother)
                }
                SanvyaButton(action: { onClose(); viewModel.reset() }) {
                    Text(S.Feedback.done)
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    // MARK: - chips

    /// A wrapping chip row. `Layout` would be tidier; this is `ViewThatFits`-free
    /// on purpose, because a fixed three-per-row grid is predictable at every
    /// Dynamic Type size and a self-measuring flow is not.
    @ViewBuilder
    private func chipFlow(
        _ options: [String],
        selected: String,
        onTap: @escaping (String) -> Void,
        label: @escaping (String) -> String
    ) -> some View {
        let rows = stride(from: 0, to: options.count, by: 3).map {
            Array(options[$0..<min($0 + 3, options.count)])
        }
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(rows.enumerated()), id: \.offset) { _, row in
                HStack(spacing: 6) {
                    ForEach(row, id: \.self) { option in
                        SanvyaChip(label(option), isActive: selected == option) { onTap(option) }
                    }
                    Spacer(minLength: 0)
                }
            }
        }
    }

    /**
     A `feedback` namespace key, resolved.

     The generated `feedbackAreaKey` / `feedbackSeverityKey` return a key rather
     than a string, for the same reason `assistantErrorKey` does: the derivation
     is shared and vector-pinned, the wording is per-locale. This is the one
     place that turns one into the other, and it is exhaustive on purpose — a
     new area in web's list fails here rather than rendering its own key on
     screen.
     */
    private func feedbackLabel(_ key: String) -> String {
        switch key {
        case "sevFatal": return S.Feedback.sevFatal
        case "sevHigh": return S.Feedback.sevHigh
        case "sevMedium": return S.Feedback.sevMedium
        case "sevLow": return S.Feedback.sevLow
        case "areaDashboard": return S.Feedback.areaDashboard
        case "areaTransactions": return S.Feedback.areaTransactions
        case "areaAccountsCards": return S.Feedback.areaAccountsCards
        case "areaBudgets": return S.Feedback.areaBudgets
        case "areaGoals": return S.Feedback.areaGoals
        case "areaInvestments": return S.Feedback.areaInvestments
        case "areaFriendsSplits": return S.Feedback.areaFriendsSplits
        case "areaSubscriptions": return S.Feedback.areaSubscriptions
        case "areaLoans": return S.Feedback.areaLoans
        case "areaAskSanvya": return S.Feedback.areaAskSanvya
        case "areaInsights": return S.Feedback.areaInsights
        case "areaSettingsBilling": return S.Feedback.areaSettingsBilling
        case "areaSyncOffline": return S.Feedback.areaSyncOffline
        case "areaOther": return S.Feedback.areaOther
        case "errNeedBug": return S.Feedback.errNeedBug
        case "errNeedSuggestion": return S.Feedback.errNeedSuggestion
        default: return S.Feedback.errSubmit
        }
    }
}

File: apps/android/app/src/main/java/com/sanvya/app/ui/feedback/FeedbackSheet.kt
package com.sanvya.app.ui.feedback

import android.content.res.Resources
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Checkbox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.sanvya.app.domain.feedback.FEEDBACK_AREAS
import com.sanvya.app.domain.feedback.FEEDBACK_SEVERITIES
import com.sanvya.app.domain.feedback.feedbackAreaKey
import com.sanvya.app.domain.feedback.feedbackSeverityKey
import com.sanvya.app.i18n.S
import com.sanvya.app.i18n.sRes
import com.sanvya.app.theme.LocalSanvyaColors
import com.sanvya.app.theme.SanvyaType
import com.sanvya.app.ui.components.Muted
import com.sanvya.app.ui.components.SanvyaButton
import com.sanvya.app.ui.components.SanvyaChip
import com.sanvya.app.ui.components.SanvyaInput
import com.sanvya.app.ui.components.SanvyaModal
import com.sanvya.app.ui.components.SanvyaText

/**
 * Send feedback -- ported from `apps/web/src/ui/BugReport.tsx`.
 *
 * The "Feedback" entry in the More sheet and the side nav was wired to a
 * closure that closed the sheet and did nothing else, on BOTH platforms: a
 * visible control that lied, and the app's only error-report channel.
 *
 * Two things are deliberately not web's:
 *
 *  * **Every string is translated.** Web's modal is hardcoded English end to
 *    end -- 30-odd strings, in an app that ships in three languages. The
 *    catalogue entries added for this port are there for web to adopt.
 *  * **The area picker is a chip grid, not a `<select>`.** Fourteen options in
 *    a native dropdown is a scroll wheel on Android and a wheel-picker sheet on
 *    iOS; chips show the whole vocabulary at once, which is what the web
 *    `<select>` effectively does on a desktop.
 *
 * What IS web's, exactly: the default-on log checkbox, the promise about what
 * is captured, the bug/suggestion split (and with it the severity row, which a
 * suggestion does not get), and the thank-you screen including the beta-tester
 * reward copy.
 *
 * Mirrors iOS's FeedbackSheet.
 */
@Composable
fun FeedbackSheet(
    open: Boolean,
    route: String,
    online: Boolean,
    onClose: () -> Unit,
    viewModel: FeedbackViewModel = viewModel(),
) {
    if (!open) return
    val res = sRes()
    val state by viewModel.state.collectAsState()
    val configuration = LocalConfiguration.current
    val viewport = "${configuration.screenWidthDp}x${configuration.screenHeightDp}"
    val userAgent = remember {
        "Android ${android.os.Build.VERSION.RELEASE}; ${android.os.Build.MANUFACTURER} ${android.os.Build.MODEL}"
    }

    SanvyaModal(
        open = true,
        onClose = { onClose(); viewModel.reset() },
        label = S.Feedback.title(res),
    ) {
        if (state.done) {
            DonePanel(
                res = res,
                isBug = state.isBug,
                onAnother = { viewModel.reset() },
                onDone = { onClose(); viewModel.reset() },
            )
            return@SanvyaModal
        }

        Column(
            modifier = Modifier.heightIn(max = (configuration.screenHeightDp * 0.7f).dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            SanvyaText(S.Feedback.title(res), SanvyaType.h2)
            Muted(S.Feedback.intro(res), style = SanvyaType.body.copy(fontSize = 12.sp))

            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                SanvyaChip(
                    S.Feedback.kindBug(res),
                    active = state.isBug,
                    onClick = { viewModel.setKind("bug") },
                    modifier = Modifier.weight(1f),
                )
                SanvyaChip(
                    S.Feedback.kindSuggestion(res),
                    active = !state.isBug,
                    onClick = { viewModel.setKind("suggestion") },
                    modifier = Modifier.weight(1f),
                )
            }

            // A suggestion has no severity -- web hides the whole row, and the
            // repository writes null, so the triage queue is not sorted by an
            // urgency nobody chose.
            if (state.isBug) {
                Muted(S.Feedback.severityLabel(res), style = SanvyaType.body.copy(fontSize = 12.sp))
                FlowRow(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                    FEEDBACK_SEVERITIES.forEach { severity ->
                        SanvyaChip(
                            feedbackLabel(res, feedbackSeverityKey(severity)),
                            active = state.severity == severity,
                            onClick = { viewModel.setSeverity(severity) },
                        )
                    }
                }
            }

            Muted(S.Feedback.areaLabel(res), style = SanvyaType.body.copy(fontSize = 12.sp))
            FlowRow(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                FEEDBACK_AREAS.forEach { area ->
                    SanvyaChip(
                        feedbackLabel(res, feedbackAreaKey(area)),
                        active = state.area == area,
                        // Tapping the chosen one again clears it: web's select
                        // has an empty first option and this has no equivalent
                        // row to offer.
                        onClick = { viewModel.setArea(if (state.area == area) "" else area) },
                    )
                }
            }

            SanvyaInput(
                value = state.title,
                onValueChange = viewModel::setTitle,
                placeholder = S.Feedback.titlePlaceholder(res),
            )
            SanvyaInput(
                value = state.description,
                onValueChange = viewModel::setDescription,
                placeholder = if (state.isBug) {
                    S.Feedback.bugPlaceholder(res)
                } else {
                    S.Feedback.suggestionPlaceholder(res)
                },
                singleLine = false,
                modifier = Modifier.heightIn(min = 96.dp),
            )

            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Checkbox(checked = state.includeLog, onCheckedChange = viewModel::setIncludeLog)
                Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                    SanvyaText(S.Feedback.includeLog(res), SanvyaType.body.copy(fontSize = 13.sp))
                    Muted(S.Feedback.includeLogHint(res), style = SanvyaType.body.copy(fontSize = 11.sp))
                }
            }

            Muted(S.Feedback.autoIncluded(res), style = SanvyaType.body.copy(fontSize = 11.sp))

            state.errorKey?.let {
                SanvyaText(
                    feedbackLabel(res, it),
                    SanvyaType.body.copy(fontSize = 13.sp),
                    color = LocalSanvyaColors.current.negative,
                )
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.End),
            ) {
                SanvyaButton(onClick = { onClose(); viewModel.reset() }, ghost = true) {
                    SanvyaText(S.Feedback.cancel(res), SanvyaType.button)
                }
                SanvyaButton(
                    onClick = {
                        viewModel.submit(
                            route = route,
                            platform = "Android",
                            userAgent = userAgent,
                            viewport = viewport,
                            online = online,
                        )
                    },
                    enabled = state.canSubmit,
                ) {
                    SanvyaText(
                        when {
                            state.busy -> S.Feedback.sending(res)
                            state.isBug -> S.Feedback.sendBug(res)
                            else -> S.Feedback.sendSuggestion(res)
                        },
                        SanvyaType.button,
                        color = Color.White,
                    )
                }
            }
        }
    }
}

/** Web's post-submit panel, reward copy and all. */
@Composable
private fun DonePanel(res: Resources, isBug: Boolean, onAnother: () -> Unit, onDone: () -> Unit) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        SanvyaText("🙏", SanvyaType.h2.copy(fontSize = 34.sp))
        SanvyaText(
            if (isBug) S.Feedback.thanksBug(res) else S.Feedback.thanksSuggestion(res),
            SanvyaType.h2,
        )
        Muted(
            if (isBug) S.Feedback.thanksBugBody(res) else S.Feedback.thanksSuggestionBody(res),
            style = SanvyaType.body.copy(fontSize = 14.sp),
        )
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            SanvyaButton(onClick = onAnother, ghost = true) {
                SanvyaText(S.Feedback.sendAnother(res), SanvyaType.button)
            }
            SanvyaButton(onClick = onDone) {
                SanvyaText(S.Feedback.done(res), SanvyaType.button, color = Color.White)
            }
        }
    }
}

/**
 * A `feedback` namespace key, resolved.
 *
 * The generated `feedbackAreaKey` / `feedbackSeverityKey` return a key rather
 * than a string, for the same reason `assistantErrorKey` does: the derivation
 * is shared and vector-pinned, the wording is per-locale. This is the one place
 * that turns one into the other, and it is exhaustive on purpose -- a new area
 * in web's list fails here rather than rendering its own key on screen.
 */
private fun feedbackLabel(res: Resources, key: String): String = when (key) {
    "sevFatal" -> S.Feedback.sevFatal(res)
    "sevHigh" -> S.Feedback.sevHigh(res)
    "sevMedium" -> S.Feedback.sevMedium(res)
    "sevLow" -> S.Feedback.sevLow(res)
    "areaDashboard" -> S.Feedback.areaDashboard(res)
    "areaTransactions" -> S.Feedback.areaTransactions(res)
    "areaAccountsCards" -> S.Feedback.areaAccountsCards(res)
    "areaBudgets" -> S.Feedback.areaBudgets(res)
    "areaGoals" -> S.Feedback.areaGoals(res)
    "areaInvestments" -> S.Feedback.areaInvestments(res)
    "areaFriendsSplits" -> S.Feedback.areaFriendsSplits(res)
    "areaSubscriptions" -> S.Feedback.areaSubscriptions(res)
    "areaLoans" -> S.Feedback.areaLoans(res)
    "areaAskSanvya" -> S.Feedback.areaAskSanvya(res)
    "areaInsights" -> S.Feedback.areaInsights(res)
    "areaSettingsBilling" -> S.Feedback.areaSettingsBilling(res)
    "areaSyncOffline" -> S.Feedback.areaSyncOffline(res)
    "areaOther" -> S.Feedback.areaOther(res)
    "errNeedBug" -> S.Feedback.errNeedBug(res)
    "errNeedSuggestion" -> S.Feedback.errNeedSuggestion(res)
    else -> S.Feedback.errSubmit(res)
}
